### PR TITLE
Add unit tests

### LINKED ISSUE #6: Add unit tests
https://api3dao.atlassian.net/jira/software/projects/BEC/boards/31?selectedIssue=BEC-642

### LINKED ISSUE #15: Configure the Github CI system to build and test the application
Like our other projects.

### PR DESCRIPTION
**Pull request by vponline**
Closes #6

### CODE REVIEW

File: src/wallet-watcher.ts [hunk 1]
```diff
@@ -21,8 +21,8 @@ import { ExtendedWalletWithMetadata, WalletStatus, WalletConfig } from './types'
  * @param walletConfig the result of parsing walletConfig.json
  * @param chainName the name of the chain
  */
-export const getProvider = async (walletConfig: WalletConfig, chainName: string) => {
-  const chainId = evm.resolveChainId(chainName);
+export const getProvider = async (walletConfig: WalletConfig, chainName: string, operations: OperationsRepository) => {
+  const chainId = evm.resolveChainId(chainName, operations as any); //TODO remove any type after operations dependency in operations-utilities is updated
```
Comment (vponline) [inline src/wallet-watcher.ts:25]: I've added `operations` as an input here for easier mocking

**Review by vponline — COMMENTED** [review]
I've added the operations fixtures from the operations repo even though most of it will likely not be needed. I'll remove the unnecessary parts with E2E tests and/or once we switch from using operations data for top ups.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  lint-compile:
+    name: Lint, Build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Clone
+        uses: actions/checkout@v2
+      - name: Setup Node
+        uses: actions/setup-node@v2
+        with:
+          node-version: '14'
```
Comment (aquarat) [inline .github/workflows/main.yml]: 14 but TARGET_NODE_VERSION is 16?

File: README.md [hunk 3]
```diff
@@ -33,3 +33,26 @@ The flag must be manually set by navigating through the AWS console (`lambda` >
...
 so. This is to prevent scenarios where dev code sends real funds to unrecoverable addresses.
+
+## Configuration
+
+### `walletConfig.json`
+
+#### `chains`
+
+- `chains.<chainId>.rpc`: The RPC provider URL.
+- `chains.<chainId>.topUpAmount`: the amount to top up in the native currency. The value should be in full token units
```
Comment (aquarat) [inline README.md]:
```suggestion
- `chains.<chainId>.topUpAmount`: the amount to top up in the native token. The value should be in full token units
```

File: README.md [hunk 4]
```diff
@@ -33,3 +33,26 @@ The flag must be manually set by navigating through the AWS console (`lambda` >
...
+- `chains.<chainId>.lowBalance`: The wallet balance value below which a top up is triggered. The value should be in full
+  token units (i.e. in ethers or matics).
+- `chains.<chainId>.globalSponsorLowBalanceWarn`: The global top up wallet balance below which an alert will be
+  triggered. The value should be in full token units (i.e. in ethers or matics).
+- `chains.<chainId>.options`: The chain specific options used to get the gas price for top up transactions.
+
+- `topUpMnemonic`: The mnemonic of the top up wallet.
+
+- `opsGenieConfig.apiKey`: The Ops Genie api key.
+- `opsGenieConfig.responders[n].team`: The Ops Genie responder type.
```
Comment (aquarat) [inline README.md]: I'm not sure if we should note that this is optional and if it isn't included/set OpsGenie will infer the team from the OpsGenie API key (we didn't realise this early on and it caused some confusion).

File: src/wallet-watcher.ts [hunk 5]
```diff
@@ -188,12 +190,13 @@ const checkAndFundWallet = async (
       return;
     }
 
-    const threshold = parseEther(globalSponsor.lowBalance);
-    if (wallet.balance.gt(threshold)) {
+    const wallerBalanceThreshold = parseEther(globalSponsor.lowBalance);
```
Comment (aquarat) [inline src/wallet-watcher.ts]: `wallerBalanceThreshold` :P

**Review by aquarat — APPROVED** [review]
Generally looks great :) 
With the idea in mind that this will need to be used by the public (and us) we should probably look at perhaps reading wallets from the config file rather than operations (a user will probably not want to fund all wallets in operations).

We'll then need a means of extracting API3 wallets from operations and printing them in JSON to the console perhaps (for our use). The format should probably be "name" and "wallet". It can derive the wallets during the conversion step. Feel free to create tasks for this and/or ask me questions.

Alternatively the wallets could be stored in their own JSON file and the converter (for our wallet monitoring) could generate this file.

**Comment by vponline** [comment]
> Generally looks great :) With the idea in mind that this will need to be used by the public (and us) we should probably look at perhaps reading wallets from the config file rather than operations (a user will probably not want to fund all wallets in operations).
> 
> We'll then need a means of extracting API3 wallets from operations and printing them in JSON to the console perhaps (for our use). The format should probably be "name" and "wallet". It can derive the wallets during the conversion step. Feel free to create tasks for this and/or ask me questions.
> 
> Alternatively the wallets could be stored in their own JSON file and the converter (for our wallet monitoring) could generate this file.

Thanks this is great input, tracking this task here: #17.

> Alternatively the wallets could be stored in their own JSON file and the converter (for our wallet monitoring) could generate this file.

Since the data section will be removed from operations, perhaps a separate JSON makes sense here. It could then just be stored on keybase with the secrets. What do you think?

File: README.md (hunk 4 above)
Comment (vponline) [reply README.md]: Yes definitely (I wasn't aware of this either).

**Pull request merged**